kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/RecordEncryptionFilterIT.java

```diff
@@ -273,6 +290,58 @@ void edekExpiry(KafkaCluster cluster, Topic topic, TestKmsFacade<?, ?, ?> testKm
         }
     }
 
+    @Test
+    void failedEncryptionRespondsWithError(KafkaCluster cluster, Topic topic, InMemoryTestKmsFacade testKmsFacade)
+            throws Exception {
+        var testKekManager = testKmsFacade.getTestKekManager();
+        testKekManager.generateKek(topic.name());
+
+        var builder = proxy(cluster);
+        // 1 second is the current minimum configurable value
+        Duration edekExpiry = Duration.ofSeconds(1);
+        builder.addToFilters(new FilterDefinitionBuilder(RecordEncryption.class.getSimpleName())
+                .withConfig("kms", testKmsFacade.getKmsServiceClass().getSimpleName())
+                .withConfig("kmsConfig", testKmsFacade.getKmsServiceConfig())
+                .withConfig("experimental", Map.of(
+                        "encryptionDekRefreshAfterWriteSeconds", edekExpiry.toSeconds(),
+                        "maxEncryptionsPerDek", 1))
+                .withConfig("selector", TemplateKekSelector.class.getSimpleName())
+                .withConfig("selectorConfig", Map.of("template", TEMPLATE_KEK_SELECTOR_PATTERN))
+                .build());
+        var messageOne = "hello world";
+        var messageTwo = "hello world2";
+        final String clientId = "producer-" + topic.name();
+        try (var tester = kroxyliciousTester(builder);
+                var producer = tester.producer(
+                        Map.of("delivery.timeout.ms", 2000,
+                                "linger.ms", 100,
+                                "request.timeout.ms", 1000,
+                                "client.id", clientId))) {
+            final AtomicReference<Exception> responseException = new AtomicReference<>();
+            producer.send(new ProducerRecord<>(topic.name(), 0, "", messageOne), (metadata, exception) -> {});
+            producer.send(new ProducerRecord<>(topic.name(), 0, "", messageTwo), (metadata, exception) -> responseException.set(exception));
+            await().until(() -> responseException.get() != null);
+            assertMetricHasCount(clientId, "connection-close-total", 0);
+            assertThat(responseException).hasValueSatisfying(
+                    actualException ->
+                            assertThat(actualException)
+                                    .isInstanceOf(TimeoutException.class)
+                                    .hasMessageStartingWith("Expiring 2 record(s) for"));
+        }
+    }
+
+    private static void assertMetricHasCount(String clientId, String metricName, double expectedValue)
```

Comment:
I wonder about adding this assertion in more tests, even after every test. If its to be used with more metrics arguably the assertion should be extracted as a function to apply...